kcmdline.c

```diff
@@ -96,13 +97,202 @@ EFI_STATUS check_cmdline(CONST CHAR8 *cmdline, UINTN cmdline_len) {
 
 	for (i=0; i < num_toks; i++) {
 		if (!is_allowed(tokens[i])) {
-			Print(L"token not allowed: %a\n", tokens[i]);
-			return EFI_SECURITY_VIOLATION;
+			UnicodeSPrint(errmsg, errmsg_len, L"token not allowed: %a\n", tokens[i]);
+			status = EFI_SECURITY_VIOLATION;
+		}
+	}
+
+out:
+
+	if (buf != NULL) {
+		FreePool(buf);
+	}
+	return status;
+}
+
+// produce the combined command line from builtin and runtime portions.
+// string lengths builtin_len and runtime_len should not include the terminating null
+// (as returned by strlen).
+EFI_STATUS get_cmdline(
+		BOOLEAN secure,
+		CONST CHAR8 *builtin, UINTN builtin_len,
+		CONST CHAR8 *runtime, UINTN runtime_len,
+		CHAR8 **cmdline, UINTN *cmdline_len,
+		CHAR16 **errmsg) {
+
+	const CHAR8 *marker = (CHAR8 *)"STUBBY_RT_CLI1";
+	const CHAR8 *namespace = (CHAR8 *)"STUBBY_RT";
+	UINTN marker_len = strlena(marker);
+	CHAR8 *p;
+	CHAR8 *part1, *part2;
```

Comment:
part1 and part2 need to be initialized to NULL so your null checks at the end of the fn before freeing can be meaningful.